app/letters/utils.py

```diff
@@ -238,3 +241,27 @@ def get_billable_units_for_letter_page_count(page_count):
     pages_per_sheet = 2
     billable_units = math.ceil(page_count / pages_per_sheet)
     return billable_units
+
+
+def adjust_daily_service_limits_for_cancelled_letters(service_id, no_of_cancelled_letters):
+    """
+    Updates the Redis values for the daily letters sent and total number of notifications sent
+    by a service. These values should be decreased if letters are cancelled.
+
+    There shouldn't be a case where these keys don't exist or where they would be changed to a
+    negative number, but this checks for those conditions anyway.
```

Comment:
I'm not sure this is true. My understanding of our rate limiting cache keys is that they don't all reset at a specific time; they last for 24 hours from created.

So if a service sends 1 letter on Monday at 1pm, it creates the cache key. Then if they send 10k letters at 10am the next day, those would be added to the existing cache key. At 1pm that cache key would expire, and they could cancel the job which would either hit a missing cache key or potentially go negative.

So a good thing you have checked those conditions.

But if you agree that this is correct then maybe good to remove or update the comment. I can't, at least, see anywhere where we clean out cache keys at a specific time.

It does make me think that, for letters at least, it probably does make sense to clear the cache key at 5:30, when we process all the letters for that day, though...

Reply:
While updating the comment I realised that I'll need to make some further changes to this PR. We use a different cache key for each day, so need to make sure we are changing the value for the day when the letters were created. For example, if I send letters yesterday they will be counted under `"<service_id>-letter-<yesterday>-count"`. But if I then try to cancel them today I will be changing `"<service_id>-letter-<today>-count"` instead.

Reply:
Yeah, I think that's where my last sentence comes in a bit.

I think for letters it may make sense to align the cache keys to the 5:30pm deadline. So either clear all the (letter limit) cache keys when that tasks starts, or do some maths on the cache key TTL to make it always expire at 5:30pm

Reply:
Even if cache keys are expired at 5.30pm (which I think it makes sense to do) this PR would still have the issue where cancelling letters sent after 5.30pm yesterday changes the wrong key - today's instead of yesterday's.

One quick solution would be to add in some logic so that we only adjust the Redis values if we're cancelling letters sent the same day. If we're cancelling letters sent yesterday there would be no point in trying to update the key from yesterday since it won't be used. I'll see how simple that would be to add.

Reply:
Ah - I didn't think of the case where you are cancelling letters from the day before (or more) - ie scheduled jobs? Good catch.

Reply:
It won't be scheduled jobs (you can only schedule texts and emails) but someone could still be cancelling a job that was created yesterday evening. I've pushed a new commit which makes sure that we only adjust the cache key if you're cancelling letters sent today.

It's still a bit weird that the letter sending period is 17.30 to 17.30 and the cache period is midnight to midnight, but that might be a bigger bit of work so could be done separately.

Reply:
Good catch. I totally missed that the cache key had the date it in - somehow 🙃